### ISSUE TITLE
Usage of IDisposable & Free() in C#

### ISSUE THREAD

**Issue by PerMalmberg** [godotengine/godot-docs#3924]
Godot 3.2.2

I've been searching/asking for a definitive answer to this question both in the [QA section](https://godotengine.org/qa/79183/idisposable-%26-free-in-c%23), [Reddit](https://www.reddit.com/r/godot/comments/ibaq0d/idisposable_free_in_c/) and on Discord, but I have yet been able to find an answer, or even anyone who dares to make a guess.

So, here's a request for documenting the intended usage of the `IDisposable` interface and `Free()`, present in both `Godot.Reference` and `Godot.Object`, and thus all sub classes of these two.

Sub classes of `Godot.Reference` are reference counted, but since the class also implements `IDisposable` it is a fair assumption that `Dispose()`should be called - but this is meant to be a one-time call which does not play well with a reference counted types.

- How are these two memory management systems intended to interact?
- Who is meant to call `Dispose()` and when?
- Who is mean to call `Free()` and when?
- Are we supposed to call both `Free()` and `Dispose()`?
- When does the engine call these methods, i.e. when does ownership of an instance pass to the engine?

**Comment by neikeq** [comment]
There are two ways to manage memory in Godot, manual or automatic, and it depends of the type you're using.
- `Reference` is automatic and it uses reference counting. A C# object of type `Reference` will hold a reference which is released when disposed or garbage collected.
- `Object` has to be freed manually. This applies to C# as well. You must call `Free`, disposing is not enough.
- `Node` can be automatic if if it has a parent (when the parent is freed it will also free the children). If the node has no parent then it has to be freed manually by calling `Free` manually as with `Object`.

When it comes to `IDisposable`, disposing a `Node` or an `Object` doesn't free the native object, so it has little impact.
In the case of `Reference` it does decrement the reference count the C# object was holding. But if you don't call dispose manually the garbage collector will take care of that, so it's not enforced.
I do recommend disposing `Reference` objects when possible though, specially if you want to control when it's released rather than relying on the unpredictable GC.

**Comment by neikeq** [comment]
We don't call `Free` automatically when disposing `Object` because we can't assume it's meant to be freed.
Luckily you don't have to deal with that most of the time if at all. For your own classes, better inherit one of the other two.

**Comment by neikeq** [comment]
I will likely make one page about lifetime of C# objects. There are also other undocumented gotchas related to this.

**Comment by PerMalmberg** [comment]
> But if you don't call dispose manually the garbage collector will take care of that, so it's not enforced.

No, the GC does not call `Dispose()`, it does however call the `finalizer`, if implemented. Dispose() is meant to be called by user-code, not the GC.

I wish I could say your answers cleared things up, @neikeq but I'm actually more confused now than before.

I get that it's tricky to get the GC to work nicely with the C++ side of the Engine, but having two memory management systems (`IDisposable` and `Free`) feels like a really bad choice. And given your answers, it seems `IDisposable`isn't actually used in the way it was intended, i.e. for releasing unmanaged resources.

I probably don't have the full picture, but would it be possible to adjust the API so that there is only one memory management system in C#? After all, people turn to C# in part because they don't want to deal with MM, now they have to deal with two.

How about using a `finalizer/Free` strategy instead of `IDisposable/Free`? That'd give the user of the engine the ability to predictably free resources, or just leave it up to the GC. There'd also be no question whether or not the item is supposed to be freed or not when Free is called (either directly or via the finalizer).

**Comment by neikeq** [comment]
>No, the GC does not call `Dispose()`, it does however call the `finalizer`, if implemented. Dispose() is meant to be called by user-code, not the GC.

I meant to say it takes care of decrementing the refcount (the finalizer calls the virtual `Dispose(disposing: false)` though).

>How about using a `finalizer/Free` strategy instead of `IDisposable/Free`? That'd give the user of the engine the ability to predictably free resources, or just leave it up to the GC. There'd also be no question whether or not the item is supposed to be freed or not when Free is called (either directly or via the finalizer).

That's definitely the solution we can't use as it means every native object accessed from C# is eventually GCed while still being accessed from native code or GDScript. We can do this with `Reference` because it also uses automatic memory management in native code. That's not the case for `Object`.

What could work is calling `Free()` when `Dispose()` is called (`disposing: true`) while not calling it from the finalizer (it may require a bit more thought for `Node` though). At least you will be safe as long as you dispose every disposable, but it's still not automatic.

I don't know, perhaps there's a solution I didn't think of that could allow the GC to manage this while not freeing what it shouldn't. I'm open for ideas.

Also keep in mind that as I said in my previous comment you rarely have to deal with objects you're required to free manually, if at all. In fact I don't know why Godot doesn't just get rid of `Object` and make `Reference` the base object of all classes. Not much would change, but that's not up-to-me.

**Comment by PerMalmberg** [comment]
> That's definitely the solution we can't use as it means every native object accessed from C# is eventually GCed while still being accessed from native code or GDScript.

So the GC and C# code is run asynchronously compared to the core engine? If not, I don't see how it's not deterministic and controlled. Shouldn't matter if you free objects from C# or gdscript; it'd all call into the C++ side of the engine. This changes nothing re. use-after-free. Am I missing something?

> Also keep in mind that as I said in my previous comment you rarely have to deal with objects you're required to free manually, if at all.

Rarely doesn't mean never, so the are cases where we're supposed to call Dispose or Free (or both?).  I imagine one case is  when removing a branch from the scene tree as that'd effectively mean you take ownership of it. This feels like a fairly common use-case.

> In fact I don't know why Godot doesn't just get rid of Object and make Reference the base object of all classes. Not much would change, but that's not up-to-me.

Performance maybe? Ref counting isn't free, although the cost ought to be really low. I like the idea.

**Comment by neikeq** [comment]
It's not much about use-after-free and more about freeing what you don't own, e.g.: a native class owning an object but then the GC freeing it. An example are singleton classes like Input, VisualServer, etc, but may not be the only cases.
